_posts/2018-07-17-its-a-lot-more-snappy.md

```diff
@@ -0,0 +1,171 @@
+---
+layout: post
+title: "\"It's A Lot More Snappy\""
+date: 2018-07-17
+categories: PWA SPA JavaScript
+author: Rick Foxcroft
+image: img/offline_logo.png
+---
+
+## _‘A real world application of PWA technology’_
+
+Recently Mechanical Rock undertook the journey of building a progressive web app (PWA) for a client here in Perth. A PWA uses modern web capabilities to deliver a native app-like experience by providing offline support and resource caching. The application we were building needed to provide users the ability to raise issues on equipment in geographically isolated areas. Due to this isolation and unreliable networks, it was important that the app could still function while offline and store data, locally on devices. We decided to build a PWA so that previously fetched data could be stored locally on the device and the application could be used without a network connection and still serve the data to the user.
+
+Whilst building the app we encountered some performance issues. Over time the app would gradually become unresponsive to the point where it would take a few seconds to register a button click.
+
+## What Was The Problem?
+
+* The app had to deal with a large dataset. One of these datasets was an array of 205509 objects. Let’s refer to these objects as Pokemon.
+* The list of Pokemon had to be used in several places throughout the app and in some places not all of the data pertaining to each Pokemon was used.
+* A JSON file of this magnitude takes a fair amount of time to download. A simulation in chrome reveals that this may take up to 3 minutes to download on  a 1mb/s link.
+
+## The Solution:
+
+* _Remove extraneous Data_ from the array and ensure this operation only happens once
+* _Update IndexedDB on a background thread_, allowing the user to progress with other aspects of the application.
+* _Improve the download speed of the data for slow networks_ by chunking requests in parallel _(beyond the scope of this post)_
+
+## 1. Remove extraneous Data
+
+```typescript
+async function transformPokemon(): Promise<void> {
+    for (let i = 0; i < this.pokemon.length; i++) {
+        const pokemon = {
+            value: this.pokemon[i].pokemonId,
+            label: `${this.pokemon[i].pokemonId} - ${this.pokemon[i].shortDescription}`
+        };
+        this.pokemonOptions.push(pokemon);
+    }
+    await localforage.setItem(this.pokemonGroup, this.pokemonOptions);
+}
+```
+
+Instead of storing the complete dataset in memory and using it in several different places. We pulled out only the data that was absolutely necessary for displaying in the UI. We also made sure that this array mapping operation only happened once every time the UPDATE_POKEMON action was dispatched to our redux store, instead of in a component render method. You'll notice, we also store a copy of this transformed array so that we don't have to retrieve it over the network the next time. Which means we can run the following when mounting our component.
```

Comment:
`Instead of storing the complete dataset in memory and using it in several different places.` Needs a comma at the end, not a full stop.